hw/ip/keymgr/doc/_index.md

```diff
@@ -244,9 +251,57 @@ Based on input from key manager control, this module selects the inputs for each
 
 ![Key Manager KMAC Interface Block Diagram](keymgr_kmac_if_diagram.svg)
 
-### Software Binding
+### Side Load Keys
+
+There are three sideload keys.
+One for AES, one for HMAC, and one for KMAC.
+When a sideload key is generated successfually through the `generate-output-hw` command, the derived data is loaded into key storage registers.
+There is a set of storage registers for each destination.
+
+The KMAC key however is further overloaded as it is the main derivation mechanism for key manager internal stage.
+The KMAC key thus has two possible outputs, one is the sideload key, and the other is internal state key.
+
+When a valid operation is called, the internal state key is sent over the KMAC key.
+During all other times, the sideloaded value is presented.
+The sideload key can be overwritten with another generate command, or cleared with entropy through {{< regref SIDELOAD_CLEAR >}}.
+
+The following diagram illustrates an example when there is no valid key in the KMAC sideload registers and an operation is called.
+During the duration of the operation, the key is valid and shows the internal key state.
+Once the operation is complete, it falls back to the sideload key state, which is invalid in this case.
+
+{{< wavejson >}}
+{signal: [
+  {name: 'u_sideload_ctrl.u_kmac_key.key_o.valid',     wave: '0................'},
+  {name: 'u_sideload_ctrl.u_kmac_key.key_o.key_share', wave: 'x................'},
+  {name: 'u_ctrl.key_o.valid',                         wave: '0................'},
+  {name: 'u_ctrl.key_o.key_share',                     wave: 'x................'},
+  {name: 'u_ctrl.op_start_i',                          wave: '0....1.....0.....'},
+  {name: 'kmac_key_o.valid',                           wave: '0....1.....0.....'},
+  {name: 'kmac_key_o.key_share*',                      wave: 'x....3.....x.....'},
+  ],
+}
+{{< /wavejson >}}
+
+The following diagram illustrates an example when there is a valid key in the KMAC sideload registers and an operation is called.
+During the duration of the operation, the key is valid and shows the internal key state.
+Once the operation is complete, it falls back to the sideload key state, which is valid and contains a different value.
+
+{{< wavejson >}}
+{signal: [
+  {name: 'u_sideload_ctrl.u_kmac_key.key_o.valid',     wave: '01...............'},
+  {name: 'u_sideload_ctrl.u_kmac_key.key_o.key_share', wave: 'x4...............'},
+  {name: 'u_ctrl.key_o.valid',                         wave: '0................'},
+  {name: 'u_ctrl.key_o.key_share',                     wave: 'x................'},
+  {name: 'u_ctrl.op_start_i',                          wave: '0....1.....0.....'},
+  {name: 'kmac_key_o.valid',                           wave: '01...............'},
+  {name: 'kmac_key_o.key_share*',                      wave: 'x4...3.....4.....'},
+  ],
```

Comment:
so what's happening in your waveform.
Is that initially there is no valid sideload key.  So that's why `u_kmac_key.key_o.valid` is 0.

You then issue a generate command, the `kmac_key_o.valid` becomes valid because the internal state key is now being sent to kmac.  Once that is done, now there is a valid key inside sideload, so that's why its valid now asserts.  Does that make sense?

Reply:
`u_kmac_key.key_o.valid` make sense to me. can you explain why `u_kmac_key.key_o.valid` is 0 in 2 cases? Thanks

Reply:
@weicaiyang i pushed an update, can you let me know if this is more clear now? (especially look at the block diagram update).

But basically, for the kmac output, there are two sources.
- sideload key 
- internal key

We only select the internal key during a valid operation (advance or generate), during all other times, the output selects the sideload key.

Now whether the sideload key is actually valid depends on whether there is a valid key in the sideload register.  If we NEVER called a generate before, then the valid may be low.  Or if we called clear, the valid may also be low.  However, if we have called a generate, then the key may just stay as valid. 

Does that help?

Reply:
Sorry, I just realized I didn't copy the signal correctly my last comment. I actually wanted to ask why `u_ctrl.key_o.valid` is 0 in 2 cases, as I see this signal is asserted when op_start_i is high in simulation. You can find it in above waveform.

Thanks for explaining the kmac output source. It really helps. Seems like the 1st waveform in sideload section, it's for invalid operation and design will output sideload key. If KMAC sideload key is generated correctly and SW doesn't clean it, is it possible we still output it when there is an invalid operation such as an operation after enter disabled state or lc_keymgr_en is turned off during an operation.

Reply:
> Sorry, I just realized I didn't copy the signal correctly my last comment. I actually wanted to ask why `u_ctrl.key_o.valid` is 0 in 2 cases, as I see this signal is asserted when op_start_i is high in simulation. You can find it in above waveform.
> 

are you referring to the period during the two `op_start_i` high regions, why `u_ctrl.key_o.valid` is 0?

> Thanks for explaining the kmac output source. It really helps. Seems like the 1st waveform in sideload section, it's for invalid operation and design will output sideload key. If KMAC sideload key is generated correctly and SW doesn't clean it, is it possible we still output it when there is an invalid operation such as an operation after enter disabled state or lc_keymgr_en is turned off during an operation.

So for the 1st waveform, there is no valid sideload key yet. So when you call an operation, assuming it is valid, it will output the internal state key, which is the yellow portion. 

And yes you are right, based on what I'm showing here, when you advanced to disable, the key output MAY still be the sideload results.  However, right now for keymgr_en, that will cause a wipe, which wipes both the internal keys and the sideload keys.  Let me update this a bit in a different PR.  What I want to trend towards will be the following:

- disable -> if software actively selects the disable operation, we will move to disable state and keep sideload key content intact, unless software also clears that via register (same as today)
- keymgr_en -> if this goes low, both the internal key and the sideload keys are wiped by hardware (same as today)
- unrecoverable error during operation -> right now this behaves like disable, but thinking about it more it might make more sense for this to behave like !keymgr_en.

Let me know what you think.

Reply:
> are you referring to the period during the two `op_start_i` high regions, why `u_ctrl.key_o.valid` is 0?

yes, it's all 0 in both waveforms. is it right? I don't pay much attention to internal signals. But since you draw them here, I try to understand more about these signals
  {name: 'u_ctrl.key_o.valid',                         wave: '0................'},
  {name: 'u_ctrl.key_o.key_share',                     wave: 'x................'},


> 
> So for the 1st waveform, there is no valid sideload key yet. So when you call an operation, assuming it is valid, it will output the internal state key, which is the yellow portion.
> 
> And yes you are right, based on what I'm showing here, when you advanced to disable, the key output MAY still be the sideload results. However, right now for keymgr_en, that will cause a wipe, which wipes both the internal keys and the sideload keys. Let me update this a bit in a different PR. What I want to trend towards will be the following:
> 
> * disable -> if software actively selects the disable operation, we will move to disable state and keep sideload key content intact, unless software also clears that via register (same as today)
> * keymgr_en -> if this goes low, both the internal key and the sideload keys are wiped by hardware (same as today)
> * unrecoverable error during operation -> right now this behaves like disable, but thinking about it more it might make more sense for this to behave like !keymgr_en.
> 
> Let me know what you think.

not sure if we need to keep the data after entering disable. Will we still use the sideload data? I feel it may be more consistent if we also wipe the key for disable. But I have very strong opinion on this since SW can also do that.

Reply:
thanks for pointing out the mistake @weicai!

regarding your second point, the main reason i don't want to wipe on disable, is because it could TECHNICALLY be a valid use case where they no longer want to be able to generate new keys, but want to use whatever has been generated before.  This would be....a really weird use case, it's almost like a terminal key use case..but it's technically possible.

Reply:
I see, thanks for the explanation.

Reply:
@tjaychen 
I'm thinking this again for the scoreboard. in above weird case, sequence is gen-hw -> enter disabled -> gen-id/sw then sideload key will appear again but kmac_key.valid will be 0. Since valid is 0, KMAC should not use this key and when we test KMAC we will make sure it doesn't use the key when valid is 0. So it's not a use case but it technically doesn't expose the valid key to KMAC.
Since we don't check data if valid is low, this doesn't affect the scoreboard.
Does this make sense?